scripts/check_if_pr_is_low_risk.py

```diff
@@ -0,0 +1,289 @@
+        dict. JSON object returned by the GitHub API v3. This is an
+        empty dictionary if the response code from the GitHub API is not
+        200.
+    """
+    request = python_utils.url_request(
+        GITHUB_API_PR_ENDPOINT % (owner, repo, pull_number),
+        None,
+        {'Accept': 'application/vnd.github.v3+json'})
+    response = python_utils.url_open(request)
+    if response.getcode() != 200:
+        return {}
+    pr = json.load(response)
+    response.close()
+    return pr
+
+
+def parse_diff(diff):
+    """Parse a PR diff into the changes made in each file.
+
+    Args:
+        diff: list(str). List of the right-stripped lines of the diff.
+
+    Returns:
+        dict(tuple(str, str), list(str)). A dictionary that maps from
+        tuples of (old filename, new filename) to lists of the diff
+        lines associated with those files. Context lines before the
+        first changed line and after the last changed line are excluded.
+    """
+    file_diffs = {}
+    old, new = '', ''
+    file_diff_started = False
+    for line in diff:
+        if line.startswith('diff --git '):
+            match = re.match(
+                r'^diff --git a/(?P<old>[\w.]+) b/(?P<new>[\w.]+)$', line)
+            old, new = match.group('old', 'new')
+            file_diffs[old, new] = []
+            file_diff_started = False
+            continue
+        if line.startswith('+++'):
+            file_diff_started = True
+            continue
+        if bool(old) and bool(new) and file_diff_started:
+            file_diffs[old, new].append(line)
+    for old, new in file_diffs:
+        lines = file_diffs[old, new]
+        i_start = -1
+        i_end = -1
+        for i, line in enumerate(lines):
+            if line.startswith(('-', '+')):
+                if i_start < 0:
+                    i_start = i
+                i_end = i
+        file_diffs[old, new] = lines[i_start:i_end + 1]
+    return file_diffs
+
+
+def check_if_pr_is_translation_pr(pr):
+    """Check if a PR is low-risk by virtue of being a translation PR.
+
+    To be a low-risk translation PR, a PR must:
+
+    * Be opened from a branch on the oppia/oppia repository.
+    * Be opened from the branch `translatewiki-prs`.
+    * Only change JSON files in `assets/i18n`.
+    * Not change the names of any files.
+
+    Args:
+        pr: dict. JSON object of PR from GitHub API.
+
+    Returns:
+        str. An empty string if the PR is a translation PR and low-risk,
+        else a message explaining why the PR is not low-risk.
+    """
+    source_repo = pr['head']['repo']['full_name']
+    if source_repo != 'oppia/oppia':
+        return 'Source repo is not oppia/oppia'
+    source_branch = pr['head']['ref']
+    if source_branch != 'translatewiki-prs':
+        return 'Source branch is not translatewiki-prs'
+    raw_diff = load_diff(pr['diff_url'])
+    if not raw_diff:
+        return 'Failed to load PR diff from GitHub API'
+    diff = parse_diff(raw_diff)
```

Comment:
If we only need the filename in the diff then isn't there any to make query to fetch only the changed file names:

Can we locally do `git diff upstearm/develop --name-only` instead of many multiple HTTP requests and parsing?

Reply:
We do need the contents of the diff. For example, in [`_check_changelog_pr_diff`](https://github.com/oppia/oppia/pull/12440/files/039334135307dc458e8467600e65653a70461319#diff-7b2ca1e3dcb75c9267b2a2c6abea409e78eb6564d680bb5ecdab85a2d8b35b84R178) we check that the change to `package.json` only updates the version number.

More generally, I prefer using the GitHub API here because we this script doesn't trust the PR author. That's the whole point of doing these checks. Therefore, I think it's better to not trust the git history we're given. Lastly, GitHub Actions does some strange things when checking out PRs. For example, try running `git --rev-parse HEAD` and you'll find that `HEAD` doesn't exist! krishnarao ran into this a couple months ago when trying to figure out whether Github actions merges with develop for us

Reply:
Sorry for the misunderstanding I was asking whether we should do the `git diff https://github.com/oppia/oppia develop` to fetch the diff (if needed)? And use git command to do lots of our tasks instead of us parsing through the diff to find the required things (as parsing can be complicated and can fail in an unexpected situation.)? wdyt?

Reply:
I don't think this will get around us parsing the diff manually. For example, we need to check that when `package.json` is updated in a changelog PR, only the version is updated. I don't think this is possible to check with just `git diff` commands. We need to check that the delta lines in the diff for `package.json` match this:

```diff
^-    'version': '[0-9]\.[0-9]\.[0-9]$'
^+    'version': '[0-9]\.[0-9]\.[0-9]$'
```

We have to parse the diff manually to check this. However, we can offload some of the parsing to `git diff`. For example, we can use `git diff --stat` to get the names of changed files and `git diff -U0 <filename>` to inspect the changes made to a particular file. Does this sound good?

Reply:
I think that would be great because that will help us use the git commands which is helpful now as well as will help us in the future for other tasks (if needed)!

For the change in version, can we do this:
- Check upstream/develop and fetch the version using JSON lib.
- Checkout feature branch and do the same as above
- Compare the version in both the branch?

Though I'm fine with manual parsing for this case because it will be a simple parsing after fetching the diff using `git diff -U0 package.json` (we can just compare the last 2 lines of the diff)

wdyt?

Reply:
I think we need to parse the diff directly because we aren't actually that interested in checking that the version numbers are valid. The real point of this check is to ensure that nothing else in `package.json` got changed. Parsing the JSON and comparing the version numbers won't catch that.

Also to clarify, we wouldn't just check the last 2 lines of the diff. We would also need to check that the diff only contains 2 lines (excluding the headers that `git diff` includes), again because the point is to make sure no other changes were made

Reply:
In that case, I think reading the JSON and comparing the JSON would be simple, right?

Psuedo also:
- [In feature branch] Create dict (pr_package) out of the package.json
- Checkout upstream/develop
- Create dict (upstream_package) out of the package.json
- pr_package_version =  pr_package[version]
- develop_package_version =  upstream_package[version]
- del  pr_package[version], del upstream_package[version]
- expect pr_package == upstream_package
- expect develop_package_version < pr_package_version

wdyt? (or should we go with parsing the diff just for packange.json?)

Reply:
I think your algorithm requires that any change to the changed file be reflected in the parsed form of the file. This doesn't quite hold for JSON since white space will be lost, but it's mostly true. However, I don't think it holds in general. For example, we also need to check the `about-page.constants.ts` file, which won't be so easy to parse (and which has lots of content like comments that a parser would strip out). I think this code will better generalize to more kinds of low-risk PRs if we require that the checkers only consider the diffs.

More generally, considering the diffs is a more conservative approach. For example, we only allow the version number in `package.json` to be updated. We don't allow adding extra spaces between the key and the value, and we don't allow moving the `version` key somewhere else. These changes wouldn't change the semantic meaning of the file, so they would be allowed by your algorithm. It's hard to see how those changes would cause problems, so maybe they should be allowed. However, I think it makes sense to be overly strict about the changes we're willing to skip CI checks for.

Reply:
> However, I think it makes sense to be overly strict about the changes we're willing to skip CI checks for.

I agree with this, as we are expecting to skip-ci for automated changes and automated changes will never add space or comments in the JSON file. In that case, I think we should use file-specific diff for expected patterns, maybe? [Suggesting using git-tool as having git command in hand for this service will be very helpful for us in future as well, but I'm fine with either i.e, fetch diff through URL and parse the full diff OR fetch changed-filenames and file-specific diff through git]